File: test/napi/lib/symbols.js
const addon = require('..');
const { assert } = require('chai');

describe('JsSymbol', function () {
    it('should return a JsSymbol with a description built with the context helper in Rust', function () {
        const sym = addon.return_js_symbol_from_context_helper();
        assert.typeOf(sym, 'symbol');
        assert.equal(sym.description, 'neon:context_helper');
    });

    it('should return a JsSymbol with a description built in Rust', function () {
        const description = 'neon:description'
        const sym = addon.return_js_symbol_with_description(description);
        assert.typeOf(sym, 'symbol');
        assert.equal(sym.description, description);
    });

    it('should return a JsSymbol without a description built in Rust', function () {
        const sym = addon.return_js_symbol();
        assert.typeOf(sym, 'symbol');
        assert.equal(sym.description, undefined);
    });

    it('should read the description property in Rust', function () {
        const sym = Symbol('neon:description');
        const description = addon.read_js_symbol_description(sym);
        assert.equal(description, 'neon:description');
    });

    it('should read an undefined description property in Rust', function () {
        const sym = Symbol();
        const description = addon.read_js_symbol_description(sym);
        assert.equal(description, undefined);
    });

    it('accepts and returns symbols', function () {
        const symDesc = Symbol('neon:description');
        const symNoDesc = Symbol();
        assert.equal(addon.accept_and_return_js_symbol(symDesc), symDesc);
        assert.equal(addon.accept_and_return_js_symbol(symNoDesc), symNoDesc);
    });
});

File: test/napi/src/js/symbols.rs
use neon::prelude::*;

pub fn return_js_symbol_from_context_helper(mut cx: FunctionContext) -> JsResult<JsSymbol> {
    Ok(cx.symbol("neon:context_helper"))
}

pub fn return_js_symbol_with_description(mut cx: FunctionContext) -> JsResult<JsSymbol> {
    let description: Handle<JsString> = cx.argument(0)?;
    Ok(JsSymbol::with_description(&mut cx, description))
}

pub fn return_js_symbol(mut cx: FunctionContext) -> JsResult<JsSymbol> {
    Ok(JsSymbol::new(&mut cx))
}

pub fn read_js_symbol_description(mut cx: FunctionContext) -> JsResult<JsValue> {
    let symbol: Handle<JsSymbol> = cx.argument(0)?;
    symbol
        .description(&mut cx)
        .map(|v| Ok(v.upcast()))
        .unwrap_or_else(|| Ok(cx.undefined().upcast()))
}

pub fn accept_and_return_js_symbol(mut cx: FunctionContext) -> JsResult<JsSymbol> {
    let sym: Handle<JsSymbol> = cx.argument(0)?;
    Ok(sym)
}
